### PR TITLE
feat: port some sources to lua

### PR DESCRIPTION
**Pull request by elianiva**
I managed to port vsnip/tags/spell. I only got [this far](https://paste.sh/2NGX8Fgi#-0t47uKjgRP6-HGCtbOneQ74) for path completion and it doesn't work so I didn't include it here.

### CODE REVIEW

File: lua/compe_tags/init.lua [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+  tags = vim.tbl_map(function(item)
+    if not vim.tbl_contains(tags, item) then
+      return item.filename
+    end
+  end, vim.fn.taglist(word))
+
+  if #tags > 10 then
+    tags = table.insert(
+      slice(tags, 1, 9), string.format("...and %d more", #slice(tags, 10))
```
Comment (elianiva) [inline lua/compe_tags/init.lua:50]: I couldn't test this one so I don't know if this work or not.

File: after/plugin/compe_tags.vim [hunk 2]
```diff
@@ -4,6 +4,7 @@ endif
 let g:loaded_compe_tags = v:true
 
 if exists('g:loaded_compe')
-  call compe#register_source('tags', compe_tags#source#create())
+  " call compe#register_source('tags', compe_tags#source#create())
+  lua require'compe'.register_source('tags', require'compe_tags')
 endif
```
Comment (hrsh7th) [inline after/plugin/compe_tags.vim:9]: Please remove comments.

File: after/plugin/compe_vsnip.vim [hunk 3]
```diff
@@ -4,6 +4,7 @@ endif
 let g:loaded_compe_vsnip = v:true
 
 if exists('g:loaded_compe')
-  call compe#register_source('vsnip', compe_vsnip#source#create())
+  " call compe#register_source('vsnip', compe_vsnip#source#create())
+  lua require'compe'.register_source('vsnip', require'compe_vsnip')
 endif
```
Comment (hrsh7th) [inline after/plugin/compe_vsnip.vim:9]: ditto

File: lua/compe_spell/init.lua [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+function Source.complete(_, context)
+  context.callback({
+    items = vim.fn.spellsuggest(context.input),
+    incomplete = true
+  })
+end
+
+return Source.new()
+
```
Comment (hrsh7th) [inline lua/compe_spell/init.lua:28]: LGTM

File: lua/compe_tags/init.lua (hunk 1 above)
Comment (hrsh7th) [reply lua/compe_tags/init.lua:50]: The below code more simple but this implementation is enough to me.

```lua
  local items = {}
  local tags = vim.fn.taglist(args.completed_item.word)
  for i, tag in ipairs(tags) do
    if 10 < i then
      table.insert(items, ('...and %d more'):format(#tags - 10))
      break
    end
    table.insert(items, tag.filename)
  end
```

File: lua/compe_tags/init.lua [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    )
+  end
+
+  context.callback(tags)
+end
+
+return Source.new()
+
```
Comment (hrsh7th) [inline lua/compe_tags/init.lua:58]: LGTM.

File: lua/compe_vsnip/init.lua [hunk 6]
```diff
@@ -0,0 +1,53 @@
+local compe = require("compe")
```
Comment (hrsh7th) [inline lua/compe_vsnip/init.lua:1]: LGTM.

File: lua/compe_tags/init.lua (hunk 1 above)
Comment (elianiva) [reply lua/compe_tags/init.lua]: that looks simpler and better, thanks!

**Comment by hrsh7th** [comment]
This PR almost looks good to me.

I think it would be good to use `for ... in ipairs` is good in Lua (In JavaScript, It will readable because `map/filter/reduce` are method.)

File: lua/compe_vsnip/init.lua [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  local items = {}
+  local snips = vim.fn['vsnip#get_complete_items'](vim.fn.bufnr('%'))
+
+  local add_user_data = function(item)
+    item.user_data = { compe = item.user_data }
+    return item
+  end
+
+  for _, v in ipairs(snips) do
+    table.insert(items, add_user_data(v))
```
Comment (elianiva) [inline lua/compe_vsnip/init.lua:29]: I think I can use `vim.tbl_map` here, one sec.

File: lua/compe_vsnip/init.lua [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+function Source.complete(_, context)
+  local items = vim.fn['vsnip#get_complete_items'](vim.fn.bufnr('%'))
+
+  local add_user_data = function(item)
+    item.user_data = { compe = item.user_data }
+    return item
+  end
+
+  vim.tbl_map(function(item) return add_user_data(item) end, items)
```
Comment (hrsh7th) [inline lua/compe_vsnip/init.lua:27]: 👍

Comment (hrsh7th) [reply lua/compe_vsnip/init.lua:27]: They both look good.

```lua
local items = vim.tbl_map(function(item)
  item.user_data = { compe = item.user_data }
  return item
end, vim.call('vsnip#get_complete_items', vim.fn.bufnr('%')))
```

```lua
local items = {}
for _, item in ipairs(vim.call('vsnip#get_complete_items', vim.fn.bufnr('%'))) do
  item.user_data = { compe = item.user_data }
  table.insert(items, item)
end
```

Comment (hrsh7th) [reply lua/compe_vsnip/init.lua:27]: Anyway, The your implementation is very good for now!

**Comment by hrsh7th** [comment]
Thank you very much!!!

**Pull request merged**